Do buildtests on our public header files with C++ as well

levitte opened the pull request:
This ensures that we don't mistakenly use C++ keywords anywhere public.

Related to #8313

levitte committed:
Do buildtests on our public header files with C++ as well

This ensures that we don't mistakenly use C++ keywords anywhere public.

Related to #8313

mspncp commented on test/build.info:
```diff
@@ -593,11 +593,19 @@ ENDIF
+
+  PROGRAMS{noinst}=buildtest_cc_$name
+  SOURCE[buildtest_cc_$name]=buildtest_$name.cc
+  GENERATE[buildtest_$name.cc]=generate_buildtest.pl $name
```
Is there a particular reason why you create extra `buildtest_*.cc` files, instead of just using the `buildtest_*.c` files? I ask, because currently the file contents are identical:

```
for f in $(find test -name '*.cc'); do echo diff $f $(dirname $f)/$(basename $f .cc).c; done
```

mspncp replied:
Ups, I had to remove an `echo` from the previous command.  ;-)

mspncp replied:
Ah forget my last question, it was stupid. Of course the compiler gets selected by the file extension.

levitte replied:
I was considering using attributes to sway the compiler choice, but that's a much bigger effort, wasn't sure it was worth it for such small files.

richsalz replied:
Does this mean A C++ compiler is required to do 'make test' now?

levitte replied:
No, it means that extra build test will be performed *if* a C++ compiler has been configured.

mspncp commented:
The xcode builds need some fixing:

```
cc1plus: warning: command line option '-std=c90' is valid for C/ObjC but not for C++ [enabled by default]
cc1plus: warning: command line option '-Wmissing-prototypes' is valid for C/ObjC but not for C++ [enabled by default]
cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++ [enabled by default]
cc1plus: warning: command line option '-Wmissing-prototypes' is valid for C/ObjC but not for C++ [enabled by default]
cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++ [enabled by default]
cc1plus: error: command line option '-std=c90' is valid for C/ObjC but not for C++ [-Werror]
cc1plus: all warnings being treated as errors
make[1]: *** [test/buildtest_cc_aes-bin-buildtest_aes_cc.o] Error 1
```

https://travis-ci.org/openssl/openssl/jobs/498095712#L3146-L3152

levitte commented:
Ah, no one has tried the boring test with Xcode then :wink:

mspncp commented:
> Ah, no one has tried the boring test with Xcode then

Doing it was probably too boring...  :laughing:

levitte commented:
Okie, `.travis.yml` adds `-std=c89`...  we could make that `-ansi` instead, that covers both C and C++ (`-ansi` is equivalent to `-std=c++98` for C++).  For the rest, I guess we'll have to make the devteam warning lists language specific.  No biggie

levitte committed:
Configuration: divide devteam flags into language specific sets

Some of the devteam flags are not for C++

levitte committed:
.travis.yml: change -std=c89 to -ansi

For C, -ansi is equivalent to -std=c90
For C++, -ansi is equivalent to -std=c++98

We also place -ansi in CPPFLAGS instead of the usual command line config,
to avoid getting it when linking (clang complains)

levitte commented:
Pushed a couple of commits that I think will clear the Travis errors

levitte commented:
Interesting...  so `clang` will be silent when `-std=c89` is given while linking, but gods forbid that `-ansi` is on that command line...

levitte commented:
CIs seem to finally be happy.

levitte commented:
Argh...  `CPPFLAGS=-ansi` didn't catch...

levitte committed:
Configure: support a few more "make variables" defaulting from env

CFLAGS, CXXFLAGS, CPPFLAGS, LDFLAGS, and LDLIBS

levitte commented:
I believe I fixed that issue.

paulidale commented on Configure:
```diff
@@ -545,18 +554,18 @@ my %user = (
     ASFLAGS     => [],
     CC          => env('CC'),
-    CFLAGS      => [],
+    CFLAGS      => [ env('CFLAGS') || () ],
```
Feature creep?

levitte replied:
Uhmmm, no?  Or how do you mean?

paulidale replied:
Adding the cflags environment read?  that's new isn't it?

levitte replied:
I can make this PR master-only, if you prefer.

paulidale replied:
I'm okay with this change going to 1.1.1.  It makes sense.

levitte replied:
Okie. The added CPPFLAGS is *needed*, and my thinking is that if I add support for that, it would be weird to leave the other "classic" variables out...

levitte commented:
Merged.

master:
8e7984e5783877c58cddc7b4e668401580ab4467 Configure: support a few more "make variables" defaulting from env
874f785988c17991051d36a0407a87b36c463a94 .travis.yml: change -std=c89 to -ansi
e373c70a3e535b560f6b6bade914a724aa975c55 Configuration: divide devteam flags into language specific sets
9f27d4bf32c0465270e1922365b21825a0f7a42a Do buildtests on our public header files with C++ as well

For 1.1.1, this requires a proper backport.  The attribute stuff added in master makes too much of a difference.
